File: src/main/java/com/fpt/service/implementations/OrderService.java
package com.fpt.service.implementations;

import com.fpt.dto.*;
import com.fpt.entity.*;
import com.fpt.form.OrderCreateRequest;
import com.fpt.form.OrderItemRequest;
import com.fpt.repository.*;
import com.fpt.service.interfaces.IBranchService;
import com.fpt.service.interfaces.IOrderService;
import com.fpt.specification.BranchSpecificationBuilder;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class OrderService implements IOrderService {

    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final OptionRepository optionRepository;
    private final CustomerRepository customerRepository;
    @Override
    @Transactional
    public OrderDTO createOrder(OrderCreateRequest request) {

        // =========================
        // 1. FIND OR CREATE CUSTOMER
        // =========================
        Customer customer = customerRepository
                .findByPhone(request.getReceiverPhone())
                .orElseGet(() -> customerRepository.save(
                        Customer.builder()
                                .phone(request.getReceiverPhone())
                                .totalOrders(0)
                                .totalSpent(0.0)
                                .build()
                ));

        // =========================
        // 2. CREATE ORDER (CHƯA TÍNH TIỀN)
        // =========================
        Order order = Order.builder()
                .status("NEW")
                .customer(customer)
                .receiverName(request.getReceiverName())
                .receiverPhone(request.getReceiverPhone())
                .deliveryAddress(request.getDeliveryAddress())
                .totalAmount(0)
                .items(new ArrayList<>())
                .build();

        order = orderRepository.save(order); // ⚠️ BẮT BUỘC SAVE TRƯỚC

        int totalAmount = 0;

        // =========================
        // 3. CREATE ORDER ITEMS
        // =========================
        for (OrderItemRequest itemReq : request.getItems()) {

            Product product = productRepository.findById(itemReq.getProductId())
                    .orElseThrow(() -> new RuntimeException("Product not found"));

            OrderItem orderItem = OrderItem.builder()
                    .order(order)
                    .product(product)
                    .quantity(itemReq.getQuantity())
                    .basePrice(product.getPrice())
                    .options(new ArrayList<>())
                    .build();

            double itemTotal = product.getPrice() * itemReq.getQuantity();

            if (itemReq.getOptionIds() != null) {
                for (Long optionId : itemReq.getOptionIds()) {

                    Option option = optionRepository.findById(optionId)
                            .orElseThrow(() -> new RuntimeException("Option not found"));

                    OrderItemOption oio = OrderItemOption.builder()
                            .orderItem(orderItem)
                            .optionId(option.getId())
                            .optionName(option.getName())
                            .optionPrice(option.getPrice())
                            .build();

                    orderItem.getOptions().add(oio);
                    itemTotal += option.getPrice();
                }
            }

            totalAmount += itemTotal;
            order.getItems().add(orderItem); // ⚠️ KHÔNG save orderItem
        }

        // =========================
        // 4. UPDATE ORDER TOTAL
        // =========================
        order.setTotalAmount(totalAmount);

        // =========================
        // 5. UPDATE CUSTOMER STATS
        // =========================
        customer.setTotalOrders(customer.getTotalOrders() + 1);
        customer.setTotalSpent(customer.getTotalSpent() + totalAmount);

        customerRepository.save(customer);

        // =========================
        // 6. SAVE ORDER (CASCADE)
        // =========================
        orderRepository.save(order);

        return mapToResponse(order);
    }

    @Override
    @Transactional(readOnly = true)
    public OrderDTO getById(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found"));

        return mapToResponse(order);
    }


    // ================= MAPPER =================
    private OrderDTO mapToResponse(Order order) {
        Customer customer = order.getCustomer();
        CustomerDTO buyerDto = null;
        if (customer != null) {
            buyerDto = CustomerDTO.builder()
                    .id(customer.getId())
                    .phone(customer.getPhone())
                    .note(customer.getNote())
                    .build();
        }
        return OrderDTO.builder()
                .id(order.getId())
                .status(order.getStatus())
                .totalAmount(order.getTotalAmount())
                .receiverName(order.getReceiverName())
                .receiverPhone(order.getReceiverPhone())
                .deliveryAddress(order.getDeliveryAddress())
                .customer(buyerDto)
                .note(order.getNote())
                .createdAt(order.getCreatedAt())
                .items(
                        order.getItems().stream().map(item ->
                                OrderItemDTO.builder()
                                        .productId(item.getProduct().getId())
                                        .productName(item.getProduct().getName())
                                        .quantity(item.getQuantity())
                                        .basePrice(item.getBasePrice())
                                        .options(
                                                item.getOptions().stream().map(opt ->
                                                        OrderItemOptionDTO.builder()
                                                                .optionId(opt.getOptionId())
                                                                .optionName(opt.getOptionName())
                                                                .optionPrice(opt.getOptionPrice())
                                                                .build()
                                                ).toList()
                                        )
                                        .build()
                        ).toList()
                )
                .build();
    }
}

File: src/modules/orders/services/order.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, DataSource } from 'typeorm';
import { Order } from '../entities/order.entity';
import { OrderItem } from '../entities/order-item.entity';
import { OrderItemOption } from '../entities/order-item-option.entity';
import { OrderCreateRequest } from '../dto/order-create.request';
import { Customer } from '../../customers/entities/customer.entity';
import { Product } from '../../products/entities/product.entity';
import { Option } from '../../products/entities/option.entity';

@Injectable()
export class OrderService {
  constructor(
    private dataSource: DataSource,
    @InjectRepository(Order)
    private orderRepository: Repository<Order>,
    @InjectRepository(Customer)
    private customerRepository: Repository<Customer>,
    @InjectRepository(Product)
    private productRepository: Repository<Product>,
    @InjectRepository(Option)
    private optionRepository: Repository<Option>,
  ) {}

  async createOrder(request: OrderCreateRequest): Promise<Order> {
    const queryRunner = this.dataSource.createQueryRunner();
    await queryRunner.connect();
    await queryRunner.startTransaction();

    try {
      // 1. Tìm hoặc tạo khách hàng
      let customer = await queryRunner.manager.findOneBy(Customer, {
        phone: request.receiverPhone,
      });

      if (!customer) {
        customer = queryRunner.manager.create(Customer, {
          phone: request.receiverPhone,
          totalOrders: 0,
          totalSpent: 0,
        });
        customer = await queryRunner.manager.save(customer);
      }

      // 2. Khởi tạo Order
      const order = queryRunner.manager.create(Order, {
        status: 'NEW',
        customer: customer,
        receiverName: request.receiverName,
        receiverPhone: request.receiverPhone,
        deliveryAddress: request.deliveryAddress,
        note: request.note,
        totalAmount: 0,
        items: [],
      });

      let totalAmount = 0;

      // 3. Xử lý từng món trong đơn
      for (const itemReq of request.items) {
        const product = await queryRunner.manager.findOneBy(Product, {
          id: itemReq.productId,
        });
        if (!product) throw new NotFoundException(`Product ${itemReq.productId} not found`);

        const orderItem = queryRunner.manager.create(OrderItem, {
          order: order,
          product: product,
          quantity: itemReq.quantity,
          basePrice: product.price,
          note: itemReq.note,
          options: [],
        });

        let itemTotal = product.price * itemReq.quantity;

        // Xử lý Options
        if (itemReq.optionIds && itemReq.optionIds.length > 0) {
          for (const optionId of itemReq.optionIds) {
            const option = await queryRunner.manager.findOneBy(Option, { id: optionId });
            if (!option) throw new NotFoundException(`Option ${optionId} not found`);

            const oio = queryRunner.manager.create(OrderItemOption, {
              orderItem: orderItem,
              optionId: option.id,
              optionName: option.name,
              optionPrice: option.price,
            });

            orderItem.options.push(oio);
            itemTotal += option.price;
          }
        }

        totalAmount += itemTotal;
        order.items.push(orderItem);
      }

      order.totalAmount = totalAmount;

      // 4. Cập nhật thống kê khách hàng
      customer.totalOrders += 1;
      customer.totalSpent += totalAmount;
      await queryRunner.manager.save(customer);

      // 5. Lưu Order (Cascade sẽ lưu cả items và options)
      const savedOrder = await queryRunner.manager.save(order);

      await queryRunner.commitTransaction();
      return savedOrder;
    } catch (err) {
      await queryRunner.rollbackTransaction();
      throw err;
    } finally {
      await queryRunner.release();
    }
  }

  async getById(id: number): Promise<Order> {
    const order = await this.orderRepository.findOne({
      where: { id },
      relations: ['customer', 'items', 'items.product', 'items.options'],
    });
    if (!order) throw new NotFoundException('Order not found');
    return order;
  }
}
